internal/server/signals.go

```diff
@@ -0,0 +1,28 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os/signal"
+	"syscall"
+
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+// RunWithSignalHandler installs a signal handler that cancels the context on SIGINT and SIGTERM, sends a SIGTERM to the
+// process group, and exits.
+func RunWithSignalHandler(ctx context.Context, run func(ctx context.Context) error) error {
+	logger := log.FromContext(ctx)
+	ctx, stop := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
```

Comment:
Doesn't the work around context cancellation happen asynchronously? Given that we immediately send SIGTERM won't it be non-deterministic as to which stuff actually sees the context cancellation?

Reply:
"run" won't return until it and all its child goroutines are complete, at which point

Reply:
What makes 'run' not return until child goroutines complete?

Reply:
The design of the function... using errgroups to track sub-tasks, etc...

Reply:
Correct, this is just the introduction of the basic cancellation.

Actually I may not be able to merge this. Turns out the only way the BuildEngine will terminate currently is by being literally killed, so yep.